lib/active_merchant/billing/gateways/nuvei.rb

```diff
@@ -0,0 +1,328 @@
+# coding: utf-8
+module ActiveMerchant #:nodoc:
+  module Billing #:nodoc:
+    class NuveiGateway < Gateway
+      self.test_url = 'https://ppp-test.safecharge.com/ppp/api/v1/'
+      self.live_url = 'https://secure.safecharge.com/ppp/api/v1/'
+      self.default_currency = 'USD'
+      self.homepage_url = 'https://www.nuvei.com/'
+      self.display_name = 'Nuvei'
+
+      def initialize(options = {})
+        requires!(options, :merchant_id, :merchant_site_id, :secret)
+        @merchant_id, @merchant_site_id, @secret = options.values_at(:merchant_id, :merchant_site_id, :secret)
+        super
+      end
+
+      def purchase(money, payment, options = {})
+        post = init_post
+        session = open_session
+        if session['sessionToken'].blank?
+          failed_session_creation(session)
+        else
+          add_session(post, session)
+          add_payment(post, money, payment, options)
+          add_device_details(post, options)
+          add_billing_address(post, options)
+          commit('payment', post, options)
+        end
+      end
+
+      def authorize(money, payment, options = {})
+        post = init_post
+        session = open_session
+        if session['sessionToken'].blank?
+          failed_session_creation(session)
+        else
+          add_session(post, session)
+          add_payment(post, money, payment, options)
+          add_device_details(post, options)
+
+          commit('initPayment', post, options)
+        end
+      end
+
+      def refund(money, authorization, options = {})
+        post = init_post(options)
+        timestamp = Time.now.utc.strftime("%Y%m%d%H%M%S")
+        add_trans_details(post, money, options, timestamp)
+        add_refund_details(post, authorization, timestamp)
+        commit('refundTransaction', post, options)
+      end
+
+      def credit(money, payment, options = {})
+        post = init_post(options)
+        timestamp = Time.now.utc.strftime("%Y%m%d%H%M%S")
+        add_trans_details(post, money, options, timestamp)
+        add_device_details(post, options)
+        add_payout_details(post, options, timestamp)
+        commit('payout', post, options)
+      end
+
+      private
+
+      # TODO: What is this for Nuvei?
```

Comment:
I have a question out to my Nuvei contact to help with this.

Reply:
This is the AVS check.
Address Verification Check result which we send to customers.